### PR TITLE
Fix subtree on Windows

### PR DESCRIPTION
**Pull request by dscho**
This fixes https://github.com/git-for-windows/git/issues/3260.

### CODE REVIEW

File: contrib/subtree/git-subtree.sh [hunk 1]
```diff
@@ -5,19 +5,27 @@
...
+if test -z "$GIT_EXEC_PATH" || {
+	test "${PATH#"${GIT_EXEC_PATH}:"}" = "$PATH" && {
+		# On Windows, PATH might be Unix-style, GIT_EXEC_PATH not
+		! type -p cygpath >/dev/null 2>&1 ||
+		test "${PATH#$(cygpath -au "$GIT_EXEC_PATH"):}" = "$PATH"
+	}
+} || ! test -f "$GIT_EXEC_PATH/git-sh-setup"
 then
+	base=${0##*/}
+	base=${0##*\\}
```
Comment (derrickstolee) [inline contrib/subtree/git-subtree.sh:17]: Why two assignments here?

Comment (landstander668) [reply contrib/subtree/git-subtree.sh:17]: I expect that the second line was meant to be:
```
base=${base##*\\}
```

Comment (dscho) [reply contrib/subtree/git-subtree.sh:17]: Yes, of course! Thank you!

**Review by derrickstolee — APPROVED** [review]

**Comment by dscho** [comment]
GitGitGadget companion: https://github.com/gitgitgadget/git/pull/978.

**Review by tranhieu0807 — APPROVED** [review]

File: contrib/subtree/git-subtree.sh [hunk 2]
```diff
@@ -5,19 +5,27 @@
 # Copyright (C) 2009 Avery Pennarun <apenwarr@gmail.com>
 #
 
-if test -z "$GIT_EXEC_PATH" || test "${PATH#"${GIT_EXEC_PATH}:"}" = "$PATH" || ! test -f "$GIT_EXEC_PATH/git-sh-setup"
+if test -z "$GIT_EXEC_PATH" || {
+	test "${PATH#"${GIT_EXEC_PATH}:"}" = "$PATH" && {
+		# On Windows, PATH might be Unix-style, GIT_EXEC_PATH not
+		! type -p cygpath >/dev/null 2>&1 ||
+		test "${PATH#$(cygpath -au "$GIT_EXEC_PATH"):}" = "$PATH"
```
Comment (LukeShu) [inline contrib/subtree/git-subtree.sh:12]: Nit: That should have a couple more `"` in it:
```sh
test "${PATH#"$(cygpath -au "$GIT_EXEC_PATH"):"}" = "$PATH"
```

**Comment by dscho** [comment]
This topic has _finally_ [advanced to upstream's `next` branch](https://github.com/gitgitgadget/git/pull/978#event-4968407645).

**Comment by r-ryabyy** [comment]
I still have issues on Windows

> $ git --version
> git version 2.37.2.windows.2

I'm using a bash and there are multiple cases where the issue reproduces.

**Case 1**
System PATH contains "C:\Program Files\Git\usr\bin;C:\Program Files\Git\mingw64\libexec\git-core;"
user PATH **doesn't contain** "C:\Program Files\Git\usr\bin;C:\Program Files\Git\mingw64\libexec\git-core;"
GIT_EXEC_PATH is not exist.

a) from bash terminal command "git subtree" results correctly
b) when i run `shell.exec('git subtree')` from npm task end up with:

> It looks like either your git installation or your
> git-subtree installation is broken.
> 
> Tips:
>  - If `git --exec-path` does not print the correct path to
>    your git install directory, then set the GIT_EXEC_PATH
>    environment variable to the correct directory.
>  - Make sure that your `git-subtree` file is either in your
>    PATH or in your git exec path (`C:\Program Files\Git\usr\bin;C:\Program Files\Git\mingw64\libexec\git-core;`).
>  - You should run git-subtree as `git subtree`,
>    not as `git-subtree`.

**Case 2**
System PATH **doesn't contain** "C:\Program Files\Git\usr\bin;C:\Program Files\Git\mingw64\libexec\git-core;"
user PATH **doesn't contain** "C:\Program Files\Git\usr\bin;C:\Program Files\Git\mingw64\libexec\git-core;"
GIT_EXEC_PATH is not exist.

a) from bash terminal command "git subtree" results correctly
b) when i run `shell.exec('git subtree')` from npm task end up with:

> It looks like either your git installation or your
> git-subtree installation is broken.
> 
> Tips:
>  - If `git --exec-path` does not print the correct path to
>    your git install directory, then set the GIT_EXEC_PATH
>    environment variable to the correct directory.
>  - Make sure that your `git-subtree` file is either in your
>    PATH or in your git exec path (`C:/Program Files/Git/mingw64/libexec/git-core`).
>  - You should run git-subtree as `git subtree`,
>    not as `git-subtree`.

**Case 3**
System PATH **doesn't contain** "C:\Program Files\Git\usr\bin;C:\Program Files\Git\mingw64\libexec\git-core;"
user PATH contains "C:\Program Files\Git\usr\bin;C:\Program Files\Git\mingw64\libexec\git-core;"
GIT_EXEC_PATH is not exist.

a) from bash terminal command "git subtree" results correctly
b) when i run `shell.exec('git subtree')` from npm task end up with:

> It looks like either your git installation or your
> git-subtree installation is broken.
> 
> Tips:
>  - If `git --exec-path` does not print the correct path to
>    your git install directory, then set the GIT_EXEC_PATH
>    environment variable to the correct directory.
>  - Make sure that your `git-subtree` file is either in your
>    PATH or in your git exec path (`C:/Program Files/Git/mingw64/libexec/git-core`).
>  - You should run git-subtree as `git subtree`,
>    not as `git-subtree`.

**Case 4**
System PATH **doesn't contain** "C:\Program Files\Git\usr\bin;C:\Program Files\Git\mingw64\libexec\git-core;"
user PATH **doesn't contain** "C:\Program Files\Git\usr\bin;C:\Program Files\Git\mingw64\libexec\git-core;"
GIT_EXEC_PATH contains "C:\Program Files\Git\usr\bin;C:\Program Files\Git\mingw64\libexec\git-core;"

a) from bash terminal command "git subtree" end up with:

> It looks like either your git installation or your
> git-subtree installation is broken.
> 
> Tips:
>  - If `git --exec-path` does not print the correct path to
>    your git install directory, then set the GIT_EXEC_PATH
>    environment variable to the correct directory.
>  - Make sure that your `git-subtree` file is either in your
>    PATH or in your git exec path (`C:\Program Files\Git\usr\bin;C:\Program Files\Git\mingw64\libexec\git-core;`).
>  - You should run git-subtree as `git subtree`,
>    not as `git-subtree`.
> 

b) when i run `shell.exec('git subtree')` from npm task end up with:

> It looks like either your git installation or your
> git-subtree installation is broken.
> 
> Tips:
>  - If `git --exec-path` does not print the correct path to
>    your git install directory, then set the GIT_EXEC_PATH
>    environment variable to the correct directory.
>  - Make sure that your `git-subtree` file is either in your
>    PATH or in your git exec path (`C:\Program Files\Git\usr\bin;C:\Program Files\Git\mingw64\libexec\git-core;`).
>  - You should run git-subtree as `git subtree`,
>    not as `git-subtree`.

**Case 5**
System PATH **doesn't contain** "C:\Program Files\Git\usr\bin;C:\Program Files\Git\mingw64\libexec\git-core;"
user PATH contains "C:\Program Files\Git\usr\bin;C:\Program Files\Git\mingw64\libexec\git-core;"
GIT_EXEC_PATH contains "C:\Program Files\Git\usr\bin;C:\Program Files\Git\mingw64\libexec\git-core;"

a) from bash terminal command "git subtree" end up with:

> It looks like either your git installation or your
> git-subtree installation is broken.
> 
> Tips:
>  - If `git --exec-path` does not print the correct path to
>    your git install directory, then set the GIT_EXEC_PATH
>    environment variable to the correct directory.
>  - Make sure that your `git-subtree` file is either in your
>    PATH or in your git exec path (`C:\Program Files\Git\usr\bin;C:\Program Files\Git\mingw64\libexec\git-core;`).
>  - You should run git-subtree as `git subtree`,
>    not as `git-subtree`.
> 

b) when i run `shell.exec('git subtree')` from npm task end up with:

> It looks like either your git installation or your
> git-subtree installation is broken.
> 
> Tips:
>  - If `git --exec-path` does not print the correct path to
>    your git install directory, then set the GIT_EXEC_PATH
>    environment variable to the correct directory.
>  - Make sure that your `git-subtree` file is either in your
>    PATH or in your git exec path (`C:\Program Files\Git\usr\bin;C:\Program Files\Git\mingw64\libexec\git-core;`).
>  - You should run git-subtree as `git subtree`,
>    not as `git-subtree`.

And the cherry on the pie - almost in all use cases Windows Powershell executes all commands correctly (both in shell and inside npm task)

Any thoughts ?

**Comment by dscho** [comment]
> b) when i run `shell.exec('git subtree')` from npm task end up with:
> 
> > It looks like either your git installation or your
> > git-subtree installation is broken.
> > Tips:
> > 
> > * If `git --exec-path` does not print the correct path to
> >   your git install directory, then set the GIT_EXEC_PATH
> >   environment variable to the correct directory.
> > * Make sure that your `git-subtree` file is either in your
> >   PATH or in your git exec path (`C:\Program Files\Git\usr\bin;C:\Program Files\Git\mingw64\libexec\git-core;`).
> > * You should run git-subtree as `git subtree`,
> >   not as `git-subtree`.

That error message comes from [here](https://github.com/git-for-windows/git/blob/v2.37.2.windows.2/contrib/subtree/git-subtree.sh#L14-L24).

If I were you, I would edit `C:\Program Files\Git\mingw64\libexec\git-core\git-subtree` and insert a `set -x` after the first line, and potentially also `echo "$GIT_EXEC_PATH"` and similar debug print statements, then try to run it again to diagnose this further, then iterate on that until I know what exactly is going wrong.

**Comment by dscho** [comment]
> System PATH contains "C:\Program Files\Git\usr\bin;C:\Program Files\Git\mingw64\libexec\git-core;"

Please note that it is recommended to add `C:\Program Files\Git\cmd` and no other directory from within `C:\Program Files\Git`, as the exact names and locations of those files are implementation details and hence subject to change. The only guarantee is that `C:\Program Files\Git\cmd\git.exe` will work.

**Pull request merged**